instrumentation/messaging/src/test/java/brave/messaging/ProducerHandlerTest.java

```diff
@@ -0,0 +1,104 @@
+package brave.messaging;
+
+import brave.Tracer;
+import brave.Tracing;
+import brave.propagation.StrictScopeDecorator;
+import brave.propagation.ThreadLocalCurrentTraceContext;
+import brave.propagation.TraceContext;
+import brave.propagation.TraceContextOrSamplingFlags;
+import java.util.ArrayList;
+import java.util.List;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+import org.mockito.junit.MockitoJUnitRunner;
+import zipkin2.Span;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+@RunWith(MockitoJUnitRunner.class)
+public class ProducerHandlerTest {
+  List<Span> spans = new ArrayList<>();
+  MessagingTracing messagingTracing;
+  @Mock MessagingAdapter adapter;
+  @Mock TraceContext.Extractor<Object> extractor;
+  @Mock TraceContext.Injector<Object> injector;
+  Object channel = new Object();
+  Object message = new Object();
+  Object carrier = new Object();
+  ProducerHandler<Object, Object, Object> handler;
+
+  @Before public void init() {
+    messagingTracing = MessagingTracing.newBuilder(
+      Tracing.newBuilder()
+        .currentTraceContext(ThreadLocalCurrentTraceContext.newBuilder()
+          .addScopeDecorator(StrictScopeDecorator.create())
+          .build())
+        .spanReporter(spans::add)
+        .build())
+      .build();
+
+    handler = ProducerHandler.create(messagingTracing, adapter, extractor, injector);
+  }
+
+  @After public void close() {
```

Comment:
probably want to make an after test that you checked all the spans. you can look at the base http instrumentation about this. "takeSpan".

Reply:
I've seen `takeSpan` used in IT by using `BlockingQueue`, but in the case of unit-tests, list is enough. When adding `messaging-tests` we could use `takeSpan`.